Table based permission sync feature

mservidio opened the pull request:
Adding ability to sync table-based permissions.

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [X] Does your submission pass tests?
2. [X] Have you lint your code locally prior to submission?

### Changes to Core Features:

* [X] Have you added an explanation of what your changes do and why you'd like us to include them?
* [X] Have you written new tests for your core changes, as applicable?
* [X] Have you successfully ran tests with your changes locally?

mservidio commented on api/v1alpha/src/lib/metaManager.js:
```diff
@@ -46,25 +48,26 @@ async function performDatasetMetadataUpdate(projectId, datasetId, accounts) {
-                    console.log(`Deleting user: ${accessType}:${accessId} from datasetId: ${datasetId}`);
-                    metadata.access.splice(i, 1);
-                    isDirty = true;
+                    const shouldHaveAccess = underscore.findWhere(accounts, { email: accessId, emailType: accessType });
```
Improvement so that we only remove users that no longer have access to the policy here, rather than always refreshing the whole access metadata.

mservidio commented on api/v1alpha/src/lib/metaManager.js:
```diff
@@ -94,107 +99,141 @@ async function performDatasetMetadataUpdate(projectId, datasetId, accounts) {
+    const tablePolicy = await bigqueryUtil.getTableIamPolicy(projectId, datasetId, tableId);
+
+    let readBinding = {};
+    if (tablePolicy && tablePolicy.bindings) {
```
Need to harden this logic still - if tablePolicy/bindings/members comes back undefined, need to handle.

mservidio commented on shared/bigqueryUtil.js:
```diff
@@ -643,69 +714,29 @@ class BigQueryUtil {
      */
     async getDatasetsByLabel(projectId, labelKey) {
         let accessTypes = ["userByEmail", "groupByEmail"];
-        if (!QUERY_FOR_LABELS) {
```
Removed unused code.

mservidio commented on shared/bigqueryUtil.js:
```diff
@@ -714,82 +745,37 @@ class BigQueryUtil {
      * @param  {} labelKey
      */
     async getTablesByLabel(projectId, datasetId, labelKey) {
-        if (!QUERY_FOR_LABELS) {
```
Removed unused code.

mservidio commented:
@phriscage When you get a chance, could you give this a preliminary review? It's not ready for merge, still have some minor touch ups, and have to add the procedure creation to the pre-requisite setup.

phriscage commented on api/v1alpha/src/admin/bq/procedure/permissionsDiff.sql:
```diff
@@ -0,0 +1,66 @@
+  SELECT
+    p.*,
+    DENSE_RANK() OVER (PARTITION BY policyId ORDER BY createdAt desc) AS rank
+  FROM `${projectId}.datashare.policy` p
```
Can this be a dynamic variable from the config for `dataset.table`? Same for line 26

mservidio commented:
@phriscage Should be good to go, I've re-organized and cleaned up a few things. Also, added support for row-level user access using table-based policies.
